packages/ui-kit/src/components/AccessTokenProvider/AccessTokenProvider.tsx

```diff
@@ -0,0 +1,128 @@
+import React, { createContext, PropsWithChildren, useCallback, useEffect, useRef, useState } from 'react'
+import { useLog } from '../Log'
+
+export interface AccessTokenContextValue {
+  /**
+   * The returned accessToken wheter from the props or from the fetchToken function
+   */
+  accessToken?: string
+  /**
+   * If true, the access token is being fetched
+   */
+  isLoading?: boolean
+  /**
+   * Function that will be called when the access token expires.
+   */
+  onExpiredToken?: () => void
+  /**
+   * If true, the access token failed to be fetched after the maximum retries
+   */
+  failedRetry?: boolean
+}
+
+const ACCESS_TOKEN_REFRESH_INTERVAL = 3300000 // 55 minutes
+const ACCESS_TOKEN_MAX_RETRIES = 3
+
+export const AccessTokenContext = createContext<AccessTokenContextValue | undefined>({})
+
+export interface AccessTokenProviderProps extends PropsWithChildren {
+  /**
+  * Function that the provider will use to fetch the access token.
+  * @returns {Promise<string>} A promise that resolves to the access token
+  */
+  fetchToken?: () => Promise<string>
+  /**
+  * If passed, the provider will ignore the `fetchToken` function and pass this access token to all the children components.
+  */
+  accessToken?: string
+}
+
+export const AccessTokenProvider = (props: AccessTokenProviderProps) => {
+  const { children, accessToken: accessTokenFromProps, fetchToken } = props
+
+  const [isLoading, setIsLoading] = useState(accessTokenFromProps == null)
+  const [fetchedToken, setFetchedToken] = useState<string | undefined>(undefined)
+  const [failedRetry, setFailedRetry] = useState(false)
+
+  const log = useLog()
+
+  const interval = useRef<NodeJS.Timeout>()
+
+  const fetch = useCallback(async () => {
+    try {
+      const token = await fetchToken()
+      log.debug('Access token fetched successfully')
+
+      setFetchedToken(token)
+
+      return token
+    } catch (error) {
+      log.error('Failed to fetch access token', error)
```

Comment:
I would also suggest introducing `error` prop, similar to Apollo's set (`loading` and `error`). So it would be easier to handle these states.

Reply:
We may also declare the `failedRetry` as a state of `error` prop and get rid of `failedRetry` itself. So, all error states can be delivered via `error` prop.